cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSICKANSink.java

```diff
@@ -342,9 +343,30 @@ public void expirateRecords(long expirationTime) throws CygnusExpiratingError {
         } // try catch
     } // truncateByTime
 
+    private class NGSICKANRowAggregator extends NGSIGenericRowAggregator {
```

Comment:
This is a good idea. Maybe on this class you can override the aggregation method to loop all the events processed by it's super, then you can add the `/ 1000` operation to the`NGSIConstants.RECV_TIME_TS` key of the LinkedHashMap.

Reply:
I don't see how to loop, it is too hard to understand what is happening.

Reply:
I agree with you. This aggregation is a bit difficult to understand. 

First of all, [here](https://github.com/telefonicaid/fiware-cygnus/blob/0baa065cbcf55bd8e4a3fdaf3d8e7c1ba48a6fd3/doc/cygnus-ngsi/user_and_programmer_guide/debugging.md)  you can find a guide to debug Cygnus with docker in order to understand better how it works.

About the implementation.

In this case (Row-storage) all events are processed by the `NGSIGenericRowAggregator.aggregate` method on each sink. Exactly [here](https://github.com/telefonicaid/fiware-cygnus/blob/6151ad42483d3ac0b72bc90e639012ad73c17fac/cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSICKANSink.java#L286) for this sink. The output of this `aggregation` method is a `LinkedHashMap`. This `LinkedHashMap` contains all events processed on the batch on the following way.

- The keys of the `LinkedHashMap` on Row aggregation are deffined as constants [here](https://github.com/telefonicaid/fiware-cygnus/blob/6c442f5ecd5fe46f2b795970c7c8e1e1f754e9ff/cygnus-ngsi-ld/src/main/java/com/telefonica/iot/cygnus/utils/NGSIConstants.java#L44).
- The values for those keys are `ArrayList` collections, wich contains all of the values processed on the batch.

For this solution, you are interested on the [RECV_TIME_TS](https://github.com/telefonicaid/fiware-cygnus/blob/6c442f5ecd5fe46f2b795970c7c8e1e1f754e9ff/cygnus-ngsi-ld/src/main/java/com/telefonica/iot/cygnus/utils/NGSIConstants.java#L44) constant. As mentioned before, a way to implement this, would be to get the key [RECV_TIME_TS](https://github.com/telefonicaid/fiware-cygnus/blob/6c442f5ecd5fe46f2b795970c7c8e1e1f754e9ff/cygnus-ngsi-ld/src/main/java/com/telefonica/iot/cygnus/utils/NGSIConstants.java#L44) from the `LinkedHashMap` after the [aggregation ends](https://github.com/telefonicaid/fiware-cygnus/blob/6151ad42483d3ac0b72bc90e639012ad73c17fac/cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSICKANSink.java#L286) and loop each one of the elements of the list to add the `/ 1000` operation. Then set this new list on the `LinkedHashMap` for the key [RECV_TIME_TS](https://github.com/telefonicaid/fiware-cygnus/blob/6c442f5ecd5fe46f2b795970c7c8e1e1f754e9ff/cygnus-ngsi-ld/src/main/java/com/telefonica/iot/cygnus/utils/NGSIConstants.java#L44).

Hope this helps.